src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -0,0 +1,323 @@
+end
+
+/-- `mul_shift ψ 0` is the trivial character. -/
+@[simp]
+lemma mul_shift_zero (ψ : add_char R R') : mul_shift ψ 0 = add_char.trivial :=
+begin
+  ext,
+  simp only [mul_shift_apply, zero_mul, of_add_zero, map_one, trivial_apply],
+end
+
+/-- An additive character is *primitive* iff all its multiplicative shifts by nonzero
+elements are nontrivial. -/
+def is_primitive (ψ : add_char R R') : Prop :=
+∀ (a : R), a ≠ 0 → is_nontrivial (mul_shift ψ a)
+
+/-- The map associating to `a : R` the multiplicative shift of `ψ` by `a`
+is injective when `ψ` is primitive. -/
+lemma to_mul_shift_inj_of_is_primitive {ψ : add_char R R'} (hψ : is_primitive ψ) :
+  function.injective ψ.mul_shift :=
+begin
+  intros a b h,
+  have h₁ : mul_shift ψ a * mul_shift ψ (- b) = mul_shift ψ b * mul_shift ψ (- b) :=
+    congr_fun (congr_arg has_mul.mul h) (mul_shift ψ (- b)),
+  simp only [mul_shift_mul, mul_shift_zero, add_right_neg] at h₁,
+  have h₂ := hψ (a + (- b)),
+  rw [h₁, is_nontrivial_iff_ne_trivial, ← sub_eq_add_neg, ne.def add_char.trivial,
+      ne_self_iff_false, imp_false, not_not] at h₂,
+  exact eq_of_sub_eq_zero h₂,
+end
+
+-- `add_comm_group.equiv_direct_sum_zmod_of_fintype`
+-- gives the structure theorem for finite abelian groups.
+-- This could be used to show that the map above is a bijection.
+-- We leave this for a later occasion.
+
+/-- When `R` is a field `F`, then a nontrivial additive character is primitive -/
+lemma is_primitive_of_is_nontrivial {F : Type u} [field F]
+ (ψ : add_char F R') (hψ : is_nontrivial ψ) : is_primitive ψ :=
+begin
+  intros a ha,
+  cases hψ with x h,
+  use (a⁻¹ * x),
+  rwa [mul_shift_spec, mul_inv_cancel_left₀ ha],
+end
+
+/-- Structure for a primitive additive character on a finite ring `R` into a cyclotomic extension
+of a field `R'`. It records which cyclotomic extension it is, the character, and the
+fact that the character is primitive. -/
+@[nolint has_inhabited_instance] -- can't prove that they always exist
+structure primitive_add_char (R : Type u) [comm_ring R] [fintype R] (R' : Type v) [field R'] :=
+(n : pnat)
+(char : add_char R (cyclotomic_field n R'))
+(prim : is_primitive char)
+
+/-!
+### Additive characters on `zmod n`
+-/
+
+variables {C : Type v} [comm_ring C]
+
+/-- We can define an additive character on `zmod n` when we have an `n`th root of unity `ζ : C`. -/
+def zmod_char (n : ℕ) [fact (0 < n)] {ζ : C} (hζ : ζ ^ n = 1) : add_char (zmod n) C :=
+{ to_fun := λ (a : multiplicative (zmod n)), ζ ^ a.to_add.val,
+  map_one' := by simp only [to_add_one, zmod.val_zero, pow_zero],
+  map_mul' := λ x y, by rw [to_add_mul, ← pow_add, zmod.val_add (to_add x) (to_add y),
+                            ← pow_eq_pow_mod _ hζ] }
+
+/-- An additive character on `zmod n` is trivial iff it takes the value `1` on `1`. -/
+lemma zmod_char_is_nontrivial_iff (n : ℕ) [fact (0 < n)] (ψ : add_char (zmod n) C) :
+  is_nontrivial ψ ↔ ψ (of_add 1) ≠ 1 :=
+begin
+  refine ⟨_, λ h, ⟨1, h⟩⟩,
+  { contrapose!,
```

Comment:
You don't need to indent this. `refine` has only given you one goal.

Reply:
OK.